### PR TITLE
[expo-clipboard] Deprecate `setString` -> `setStringAsync`

### PR DESCRIPTION
**Pull request by barthap**
# Why

- Resolves ENG-4141
- https://github.com/expo/expo/pull/16269#discussion_r805756462
- There was [a discussion](https://github.com/expo/expo/pull/11227#discussion_r537262204) sometime ago that we can deprecate it if we want to

# How

Updated:
- JS
   - On native, `setString` delegates to `setStringAsync` without awaiting - maintains old behavior
   - On web, `setString` calls old synchronous implementation directly, whereas `setStringAsync` calls it via `Promise.resolve()` to keep maximum backwards compatibility
- JS unit tests
- iOS native name
- Android native name
- NCL
- Docs
- TODO: Changelog

# Test Plan

NCL on all platforms, unit tests, ran docs locally

### CODE REVIEW

**Review by lukmccall — COMMENTED** [review]
Looks good 😉 Just one thing ;)

File: packages/expo-clipboard/src/Clipboard.ts [hunk 1]
```diff
@@ -33,14 +33,31 @@ export async function getStringAsync(): Promise<string> {
  *
  * @param text The string to save to the clipboard.
  *
+ * @returns On web, this returns a promise that fulfills to a boolean value indicating whether or not
+ * the string was saved to the user's clipboard. On iOS and Android, an empty promise is returned.
+ */
+export async function setStringAsync(text: string): Promise<boolean | void> {
```
[the remark this answers is not in the document]
Comment (bbarthec) [reply packages/expo-clipboard/src/Clipboard.ts:39]: I like that idea 👍 
Is it possible that Android or iOS would reject setting a string in the clipboard? 🤔

**Review by bbarthec — COMMENTED** [review]
Looking good, reply to suggestions and unify platforms as much as possible and we're ready to ship it 🚢 👍

File: packages/expo-clipboard/src/Clipboard.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (barthap) [reply packages/expo-clipboard/src/Clipboard.ts:39]: Ok, I'll do this and then merge 💪

> Is it possible that Android or iOS would reject setting a string in the clipboard?

Nope, it's just simple assignment. The only rejection cause would be when the clipboard is for some reason unavailable on the OS.

**Review by lukmccall — APPROVED** [review]

File: packages/expo-clipboard/CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,8 @@
 
 ### 🛠 Breaking changes
 
+- Deprecated `setString`. Use `setStringAsync` instead. ([#16320](https://github.com/expo/expo/pull/16320) by [@barthap](https://github.com/barthap))
```
Comment (tsapeta) [inline packages/expo-clipboard/CHANGELOG.md:7]: Deprecating is imho not a breaking change — as a result of this change users don't have to modify their code yet. You could use another changelog category, like `⚠️ Notices`

Comment (barthap) [reply packages/expo-clipboard/CHANGELOG.md:7]: You're right, I had similar thoughts, but I checked that deprecations in other changelogs were put in the `Breaking changes` section.

But I can move that if you wish 😁

Comment (tsapeta) [reply packages/expo-clipboard/CHANGELOG.md:7]: If they are put there, it's wrong 😅 Refer to [Updating Changelogs](https://github.com/expo/expo/blob/main/guides/contributing/Updating%20Changelogs.md#categories) guide.

Comment (barthap) [reply packages/expo-clipboard/CHANGELOG.md:7]: Great 😁  I changed this in https://github.com/expo/expo/commit/1ca5a5daab190970e11873e18adea86c2e90e2e9

**Pull request merged**